code/datums/components/acid.dm

```diff
@@ -0,0 +1,211 @@
+		return COMPONENT_INCOMPATIBLE // Not enough acid.
+	if(!isatom(parent))
+		stack_trace("Acid component added to [parent] ([parent?.type]) which is not a /atom subtype.")
+		return COMPONENT_INCOMPATIBLE // Incompatible type. TODO: Rework take_damage to the atom level and move this there.
+
+	if(isobj(parent))
+		var/obj/parent_object = parent
+		if(parent_object.resistance_flags & UNACIDABLE) // The parent object cannot have acid.
+			stack_trace("Acid component added to unacidable object [parent].")
+			return COMPONENT_INCOMPATIBLE
+
+		max_volume = OBJ_ACID_VOLUME_MAX
+		process_effect = CALLBACK(src, .proc/process_obj, parent)
+	else if(isliving(parent))
+		max_volume = MOB_ACID_VOLUME_MAX
+		process_effect = CALLBACK(src, .proc/process_mob, parent)
+	else if(isturf(parent))
+		max_volume = TURF_ACID_VOLUME_MAX
+		process_effect = CALLBACK(src, .proc/process_turf, parent)
+
+	acid_power = _acid_power
+	set_volume(_acid_volume)
+
+	var/atom/parent_atom = parent
+	RegisterSignal(parent, COMSIG_ATOM_UPDATE_OVERLAYS, .proc/on_update_overlays)
+	parent_atom.update_icon()
+	sizzle = new(list(parent), TRUE)
+	START_PROCESSING(SSacid, src)
+
+/datum/component/acid/Destroy(force, silent)
+	STOP_PROCESSING(SSacid, src)
+	QDEL_NULL(sizzle)
+	QDEL_NULL(process_effect)
+	UnregisterSignal(parent, COMSIG_ATOM_UPDATE_OVERLAYS)
+	if(parent && !QDELING(parent))
+		var/atom/parent_atom = parent
+		parent_atom.update_icon()
+	return ..()
+
+/datum/component/acid/RegisterWithParent()
+	RegisterSignal(parent, COMSIG_PARENT_EXAMINE, .proc/on_examine)
+	RegisterSignal(parent, COMSIG_COMPONENT_CLEAN_ACT, .proc/on_clean)
+	RegisterSignal(parent, COMSIG_ATOM_ATTACK_HAND, .proc/on_attack_hand)
+	RegisterSignal(parent, COMSIG_ATOM_EXPOSE_REAGENT, .proc/on_expose_reagent)
+	if(isturf(parent))
+		RegisterSignal(parent, COMSIG_MOVABLE_CROSSED, .proc/on_crossed)
+
+/datum/component/acid/UnregisterFromParent()
+	UnregisterSignal(parent, list(
+		COMSIG_PARENT_EXAMINE,
+		COMSIG_COMPONENT_CLEAN_ACT,
+		COMSIG_ATOM_ATTACK_HAND,
+		COMSIG_ATOM_EXPOSE_REAGENT))
+
+	if(isturf(parent))
+		UnregisterSignal(parent, COMSIG_MOVABLE_CROSSED)
+
+/// Averages corrosive power and sums volume.
+/datum/component/acid/InheritComponent(datum/component/C, i_am_original, _acid_power, _acid_volume)
+	acid_power = ((acid_power * acid_volume) + (_acid_power * _acid_volume)) / (acid_volume + _acid_volume)
+	set_volume(acid_volume + _acid_volume)
+
+/// Sets the acid volume to a new value. Limits the acid volume by the amount allowed to exist on the parent atom.
+/datum/component/acid/proc/set_volume(new_volume)
+	acid_volume = clamp(new_volume, 0, max_volume)
+	if(!acid_volume)
+		qdel(src)
+
+
+/// Handles the slow corrosion of the parent [/atom].
+/datum/component/acid/process()
+	process_effect?.InvokeAsync()
+	set_volume(acid_volume - (ACID_DECAY_BASE + (ACID_DECAY_SCALING*round(sqrt(acid_volume)))))
```

Comment:
`process_turf()` calls `set_volume()` and `process_obj()` calls `take_damage()`, meaning by the time it reaches this `set_volume()` call, `src` might be qdeleting, and this will make it try to qdelete twice.
```suggestion
	process_effect?.InvokeAsync()
	if(QDELETING(src))
		return
	set_volume(acid_volume - (ACID_DECAY_BASE + (ACID_DECAY_SCALING*round(sqrt(acid_volume)))))
```